hibernate-core/src/main/java/org/hibernate/query/sqm/sql/BaseSqmToSqlAstConverter.java

```diff
@@ -5985,7 +5985,7 @@ else if ( paramSqmType instanceof SingularAttribute ) {
 						return resolveEntityPersister( (EntityDomainType<?>) type );
 					}
 				}
-				else if ( inferredMapping != null ) {
+				if ( inferredMapping != null ) {
```

Comment:
This change, along with the changes in `TypecheckUtil`, is not strictly needed to fix this issue. I believe this is more correct when resolving a parameter's type from a `SingularAttribute` that is not entity typed.